### PR TITLE
add tf-gridnet

### PR DESCRIPTION
**Pull request by zqwang7**
Code of TF-GridNet proposed in

[1] Z.-Q. Wang, S. Cornell, S. Choi, Y. Lee, B.-Y. Kim, and S. Watanabe,
    "TF-GridNet: Integrating Full- and Sub-Band Modeling for Speech Separation",
    in arXiv preprint arXiv:2211.12433, 2022.
[2] Z.-Q. Wang, S. Cornell, S. Choi, Y. Lee, B.-Y. Kim, and S. Watanabe,
    "TF-GridNet: Making Time-Frequency Domain Models Great Again for Monaural Speaker Separation",
    in arXiv preprint arXiv:2209.03952, 2022.

The SI-SDR result of using D=48, I=4, J=1, H=192 (and 16 ms window and 8 ms hop sizes) obtains 22.8 dB SI-SNR on WSJ0-2mix. It is close to the 23.2 dB (obtained using 32 ms window and 8 ms hop sizes) reported in the jounral submission. We think the result is reasonably close, considering that, when using 4 s chunk_length and the default SI-SNR loss for training, the current chunking mechanism in ESPNet-SE discards ~15% training examples and some trailing segments (each of which can be up to 2-second long).

### CODE REVIEW

**Comment by sw005320** [comment]
Please fix the CI issue
https://github.com/espnet/espnet/actions/runs/3895640525/jobs/6657294336#step:8:10

**Comment by Emrys365** [comment]
Could you add some unit tests to `tests/espnet2/enh`?

**Comment by Emrys365** [comment]
It would be great if you also share the trained model on HuggingFace.

**Comment by sw005320** [comment]
Any update, @zqwang7?
We're very keen to use TF gridnet in the other applications.

**Comment by sw005320** [comment]
FYI, https://github.com/espnet/espnet/actions/runs/4000248137/jobs/6865197032#step:8:9966

**Comment by zqwang7** [comment]
> FYI, https://github.com/espnet/espnet/actions/runs/4000248137/jobs/6865197032#step:8:9966

In the most recent commits, "[fix isort](https://github.com/espnet/espnet/pull/4864/commits/2d7a812c1020871de0de13745a96e65a72648a5e)" changes

from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainL1, FrequencyDomainMSE

to

from espnet2.enh.loss.criterions.tf_domain import (FrequencyDomainL1,
                                                   FrequencyDomainMSE)

But then, there is a code format error, and I need to run black ([run black](https://github.com/espnet/espnet/pull/4864/commits/ddf4a07bd3fc370250a624bc8eb58b7692e2a933)), and black changes 

from espnet2.enh.loss.criterions.tf_domain import (FrequencyDomainL1,
                                                   FrequencyDomainMSE)

to

from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainL1, FrequencyDomainMSE

I have no ideas how to deal with this....

**Comment by zqwang7** [comment]
> > FYI, https://github.com/espnet/espnet/actions/runs/4000248137/jobs/6865197032#step:8:9966
> 
> In the most recent commits, "[fix isort](https://github.com/espnet/espnet/pull/4864/commits/2d7a812c1020871de0de13745a96e65a72648a5e)" changes
> 
> from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainL1, FrequencyDomainMSE
> 
> to
> 
> from espnet2.enh.loss.criterions.tf_domain import (FrequencyDomainL1, FrequencyDomainMSE)
> 
> But then, there is a code format error, and I need to run black ([run black](https://github.com/espnet/espnet/pull/4864/commits/ddf4a07bd3fc370250a624bc8eb58b7692e2a933)), and black changes
> 
> from espnet2.enh.loss.criterions.tf_domain import (FrequencyDomainL1, FrequencyDomainMSE)
> 
> to
> 
> from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainL1, FrequencyDomainMSE
> 
> I have no ideas how to deal with this....

Would changing

from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainL1, FrequencyDomainMSE

to 

from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainL1
from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainMSE

solve the problem?

**Comment by Emrys365** [comment]
It seems a version issue. Could you update both packages to the latest version and retry?

File: espnet2/enh/separator/tfgridnet_separator.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+from torch.nn import init
+import torch.nn as nn
+import torch.nn.functional as F
+
+from espnet2.enh.decoder.stft_decoder import STFTDecoder
+from espnet2.enh.encoder.stft_encoder import STFTEncoder
+from espnet2.enh.separator.abs_separator import AbsSeparator
+from espnet2.torch_utils.get_layer_from_string import get_layer
+
+from packaging.version import parse as V
```
Comment (Emrys365) [inline espnet2/enh/separator/tfgridnet_separator.py:16]: Could you move this line to line 10 and run `isort` to fix the importing order?

> We're following PEP8's style for the arranging the imports.

File: espnet2/enh/separator/tfgridnet_separator.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+        batch = self.deconv(batch)  # [B, n_srcs*2, T, F]
+
+        batch = batch.view([n_batch, self.n_srcs, 2, n_frames, n_freqs])
+        batch = torch.complex(batch[:, :, 0], batch[:, :, 1])  # [B, n_srcs, T, F]
+
+        batch = self.dec(
+            batch.view([n_batch * self.num_spk, n_frames, n_freqs]), ilens
+        )[
+            0
+        ]  # [B, n_srcs, -1]
```
Comment (Emrys365) [inline espnet2/enh/separator/tfgridnet_separator.py:164]:
```suggestion
        batch = self.dec(batch.view(-1, n_frames, n_freqs), ilens)[0] # [B, n_srcs, -1]
```

File: espnet2/enh/separator/tfgridnet_separator.py [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+        emb_ks,
+        emb_hs,
+        n_freqs,
+        hidden_channels,
+        n_head=4,
+        approx_qk_dim=512,
+        activation="prelu",
+        eps=1e-5,
+    ):
+        super(GridNetBlock, self).__init__()
```
Comment (Emrys365) [inline espnet2/enh/separator/tfgridnet_separator.py:202]:
```suggestion
        super().__init__()
```

File: espnet2/enh/separator/tfgridnet_separator.py [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+        """
+        B, C, old_T, old_Q = x.shape
+        T = (
+            math.ceil((old_T - self.emb_ks) * 1.0 / self.emb_hs) * self.emb_hs
+            + self.emb_ks
+        )
+        Q = (
+            math.ceil((old_Q - self.emb_ks) * 1.0 / self.emb_hs) * self.emb_hs
+            + self.emb_ks
+        )
```
Comment (Emrys365) [inline espnet2/enh/separator/tfgridnet_separator.py:278]: `* 1.0` can be removed as the output of division `/` is a floating-point number in Python3.

File: espnet2/enh/separator/tfgridnet_separator.py [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+        )  # [B, C, T, Q])
+        batch = self["attn_concat_proj"](batch)  # [B, C, T, Q])
+
+        out = batch + inter_rnn
+        return out
+
+
+class LayerNormalization4D(nn.Module):
+    def __init__(self, input_dimension, eps=1e-5):
+        super(LayerNormalization4D, self).__init__()
```
Comment (Emrys365) [inline espnet2/enh/separator/tfgridnet_separator.py:359]:
```suggestion
        super().__init__()
```

File: espnet2/enh/separator/tfgridnet_separator.py [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+        std_ = torch.sqrt(
+            x.var(dim=stat_dim, unbiased=False, keepdim=True) + self.eps
+        )  # [B,1,T,F]
+        x_hat = ((x - mu_) / std_) * self.gamma + self.beta
+        return x_hat
+
+
+class LayerNormalization4DCF(nn.Module):
+    def __init__(self, input_dimension, eps=1e-5):
+        super(LayerNormalization4DCF, self).__init__()
```
Comment (Emrys365) [inline espnet2/enh/separator/tfgridnet_separator.py:383]:
```suggestion
        super().__init__()
```

File: espnet2/enh/separator/tfgridnet_separator.py [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+        self.beta = Parameter(torch.Tensor(*param_size).to(torch.float32))
+        init.ones_(self.gamma)
+        init.zeros_(self.beta)
+        self.eps = eps
+
+    def forward(self, x):
+        if x.ndim == 4:
+            stat_dim = (1, 3)
+        else:
+            raise
```
Comment (Emrys365) [inline espnet2/enh/separator/tfgridnet_separator.py:396]:
```suggestion
            raise ValueError("Expect x to have 4 dimensions, but got {}".format(x.ndim))
```

**Comment by sw005320** [comment]
See https://github.com/espnet/espnet/actions/runs/4023093431/jobs/6913553984#step:8:9974
Is it the same issue that you mentioned before in https://github.com/espnet/espnet/pull/4864#issuecomment-1402775794?

**Comment by sw005320** [comment]
almost there!
https://github.com/espnet/espnet/actions/runs/4032631808/jobs/6935565456#step:8:9994

File: espnet2/enh/separator/tfgridnet_separator.py [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+                ),
+            )
+            self.add_module(
+                "attn_conv_V_%d" % ii,
+                nn.Sequential(
+                    nn.Conv2d(emb_dim, emb_dim // n_head, 1),
+                    get_layer(activation)(),
+                    LayerNormalization4DCF((emb_dim // n_head, n_freqs), eps=eps),
+                ),
+            )
```
Comment (quancs) [inline espnet2/enh/separator/tfgridnet_separator.py:245]: Hello, @zqwang7. I tried the code in this pull request, but it seems the code is slow. I guess these modules are the bottleneck. And merging these modules can largely increase the training speed.
